File: experiments/runs/2026-09-16_08-12-47_claim-office-example-mapping_exact-sol-v1.6-test-list-dimensions-cc_opus-5-no-thinking/src/customer-modifiers.ts
import type { Customer } from "./customer.js";

/** Policy-wide modifiers, charged on the policy base premium. */
const LOYALTY_DISCOUNT_PERCENT = 20;
const FIRST_INSURANCE_SURCHARGE_PERCENT = 10;
const FOLLOW_UP_CONTRACT_DISCOUNT_PERCENT = 15;
const LOYAL_CUSTOMER_YEARS = 2;

function isLongStanding(customer: Customer): boolean {
  return customer.yearsWithMHPCO >= LOYAL_CUSTOMER_YEARS;
}

function isFollowUpContract(previousContractCount: number): boolean {
  return previousContractCount > 0;
}

export function policyModifierPercent(customer: Customer, previousContractCount: number): number {
  const loyaltyPercent = isLongStanding(customer) ? -LOYALTY_DISCOUNT_PERCENT : 0;
  const followUpPercent = isFollowUpContract(previousContractCount)
    ? -FOLLOW_UP_CONTRACT_DISCOUNT_PERCENT
    : 0;
  return FIRST_INSURANCE_SURCHARGE_PERCENT + loyaltyPercent + followUpPercent;
}

File: experiments/runs/2026-09-20_07-33-00_claim-office-java-example-mapping_exact-ptdd-v1-cc_opus-5-no-thinking/src/main/java/CustomerModifiers.java
/**
 * The MHPCO's policy-wide modifiers, which depend on the customer's history
 * with the office and apply to the policy base premium.
 */
final class CustomerModifiers {

    private static final double FIRST_INSURANCE_SURCHARGE = 0.10;
    private static final double LOYALTY_DISCOUNT = 0.20;
    private static final int LOYALTY_YEARS = 2;
    private static final double FOLLOW_UP_CONTRACT_DISCOUNT = 0.15;

    private final Customer customer;
    private int contractsIssued;

    CustomerModifiers(Customer customer) {
        this.customer = customer;
    }

    /** The net rate by which the customer's history adjusts a policy base premium. */
    double netRate() {
        double rate = FIRST_INSURANCE_SURCHARGE;
        if (isLongStanding()) {
            rate -= LOYALTY_DISCOUNT;
        }
        if (contractsIssued > 0) {
            rate -= FOLLOW_UP_CONTRACT_DISCOUNT;
        }
        return rate;
    }

    void recordContract() {
        contractsIssued++;
    }

    private boolean isLongStanding() {
        return customer.yearsWithMHPCO() >= LOYALTY_YEARS;
    }
}
